component/distributed_cache/distributed_cache.go

```diff
@@ -356,68 +361,265 @@ func (dc *DistributedCache) StreamDir(options internal.StreamDirOptions) ([]*int
+			return nil, err
+		}
+	} else if isDcachePath {
+		log.Debug("DistributedCache::OpenFile : Path is having Dcache subcomponent, path : %s", options.Name)
+		options.Name = rawPath
+		dcFile, err = fm.OpenDcacheFile(options.Name)
+		if err != nil {
+			log.Err("DistributedCache::OpenFile : File Open failed with err : %s Dcache subcomponent, path : %s", err.Error(), options.Name)
+			return nil, err
+		}
+	} else {
+		common.Assert(options.Name == rawPath)
+		dcFile, err = fm.OpenDcacheFile(rawPath)
+		if err != nil {
+			log.Err("DistributedCache::OpenFile : File Open failed with err : %s Dcache subcomponent, path : %s", err.Error(), options.Name)
+			return nil, err
+		}
+		handle, err = dc.NextComponent().OpenFile(options)
+		if err != nil {
+			log.Err("DistributedCache::OpenFile : File Open failed with err : %s Azure subcomponent, path : %s", err.Error(), options.Name)
+			return nil, err
+		}
+	}
+
+	if handle == nil {
+		handle = handlemap.NewHandle(options.Name)
+	}
+
+	// Set the respective filesystems that this path can access
+	if isAzurePath {
+		handle.SetFsAzure()
+	} else if isDcachePath {
+		handle.SetFsDcache()
+	} else {
+		handle.SetFsDefault()
+	}
+
+	// Set Dcache file inside the handle
+	handle.IFObj = dcFile
+
+	return handle, nil
 }
 
 func (dc *DistributedCache) ReadInBuffer(options internal.ReadInBufferOptions) (int, error) {
-	return 0, syscall.ENOTSUP
+	log.Err("DistributedCache::ReadInBuffer : ReadInBuffer, offset : %d, buf size : %d, file : %s",
+		options.Offset, len(options.Data), options.Handle.Path)
+
+	var dcacheErr, azureErr error
+	var bytesRead int
+	if options.Handle.IsFsDcache() {
+		common.Assert(options.Handle.IFObj != nil)
+		dcFile := options.Handle.IFObj.(*fm.DcacheFile)
+		bytesRead, dcacheErr = dcFile.ReadFile(options.Offset, options.Data)
+		if dcacheErr == nil || dcacheErr == io.EOF {
+			return bytesRead, dcacheErr
+		}
+		common.Assert(bytesRead == 0)
+		log.Err("DistributedCache::ReadInBuffer : Failed to read the file from the Dcache, offset : %d, file : %s",
+			options.Offset, options.Handle.Path)
+		// Let's try to read from the azure, if the handle has the access
+	}
+	if options.Handle.IsFsAzure() {
+		bytesRead, azureErr = dc.NextComponent().ReadInBuffer(options)
+		if azureErr == nil || azureErr == io.EOF {
+			return bytesRead, azureErr
+		}
+		log.Err("DistributedCache::ReadInBuffer : Failed to read the file from the Azure, offset : %d, file : %s",
+			options.Offset, options.Handle.Path)
+	}
+	err := errors.Join(dcacheErr, azureErr)
+	common.Assert(err != nil)
+	return 0, err
 }
 
 func (dc *DistributedCache) WriteFile(options internal.WriteFileOptions) (int, error) {
-	return 0, syscall.ENOTSUP
+	log.Err("DistributedCache::WriteFile : WriteFile, offset : %d, buf size : %d, file : %s",
+		options.Offset, len(options.Data), options.Handle.Path)
+
+	if !options.Handle.Flags.IsSet(handlemap.HandleFlagDcacheAllowWrites) {
+		return 0, syscall.EIO
+	}
```

Comment:
we should assert for this?
since we would have failed OpenFile() for dcache files with write flags, fuse shouldn't call write for such an handle.
Also, if the write is for an azure file, we shouldn't disallow.